client/src/main/scala/com/redi/scalademo/presentation/FormListener.scala

```diff
@@ -55,80 +49,96 @@ class FormListener(
-      val name: String = formControlData(Name)
-      val value: String = formControlData(Value)
-      if (shouldValidate) {
-        val isValid: Boolean = numericStringValidator.isValidNumber(value)
-        if (!isValid) {
-          allFormValuesAreValid = false
-          formElement.find(elementNamed(name)).addClass(DangerClass)
-        }
-      }
-      name match {
-        case Augend ⇒
-          augend = value
-        case Addend ⇒
-          addend = value
+      val name: String = formControlData("name")
+      val value: String = formControlData("value")
+      val isValid: Boolean = numericStringValidator.isValidNumber(value)
+      if (!isValid) {
+        invalidElements += formElement.find(elementNamed(name))
       }
     }
+    invalidElements
+  }
+
+  private def calculableFormData: js.Array[js.Dictionary[String]] = {
+    serializeFormData(augendElement) ++ serializeFormData(addendElement)
+  }
+
+  private def serializeFormData(element: JQuery): js.Array[js.Dictionary[String]] = {
+    element.serializeArray().asInstanceOf[js.Array[js.Dictionary[String]]]
   }
 
   private def invokeAddition(): Unit = {
-    if (allFormValuesAreValid) {
-      val futureResult: Future[ValidatedResult[BigDecimal]] =
-        client[Calculator].add(augend, addend).call()
-      for (result: ValidatedResult[BigDecimal] ← futureResult) {
-        result match {
-          case Left(validationFailures) ⇒
-            for (validationFailure: ValidationFailure ← validationFailures) {
-              formElement
-                .find(elementNamed(validationFailure.formControlName))
-                .addClass(DangerClass)
-            }
-          case Right(summand: BigDecimal) ⇒
-            summandElement
-              .value(summand.toString)
-              .addClass(InfoClass)
-        }
+    val futureResult: Future[ValidatedResult[BigDecimal]] =
+      client[Calculator].add(augendValue, addendValue).call()
+    for (result: ValidatedResult[BigDecimal] ← futureResult) {
+      result match {
+        case Left(validationFailures) ⇒
+          for (validationFailure: ValidationFailure ← validationFailures) {
+            formElement
+              .find(elementNamed(validationFailure.formControlName))
+              .addClass(DangerClass)
+          }
+        case Right(summand: BigDecimal) ⇒
+          summandElement
+            .value(summand.toString)
+            .addClass(InfoClass)
       }
     }
   }
 
+  private def augendValue: String = {
+    augendElement.value().toString
+  }
+
+  private def addendValue: String = {
+    addendElement.value().toString
+  }
+
 }
 
 object FormListener {
 
   private val InfoClass = "is-info"
   private val DangerClass = "is-danger"
 
-  private val Validate = "validate"
-  private val Augend = "augend"
-  private val Addend = "addend"
-  private val Summand = "summand"
-
-  private val Name = "name"
-  private val Value = "value"
```

Comment:
To demo smoothly and minimize going back-and-forth in the code, I removed constants that are only used once.